rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,320 @@
+  ROSBAG2_CPP_LOG_INFO_STREAM("Extracting metadata from database(s)");
+  for (const auto & f_ : files) {
+    ROSBAG2_CPP_LOG_INFO_STREAM("Extracting from file: " + f_.string());
+
+    // Set up reader
+    rosbag2_storage::StorageOptions temp_so = {
+      f_.string(),  // uri
+      storage_options.storage_id,  // storage_id
+      storage_options.max_bagfile_size,  // max_bagfile_size
+      storage_options.max_bagfile_duration,  // max_bagfile_duration
+      storage_options.max_cache_size,  // max_cache_size
+      storage_options.storage_config_uri  // storage_config_uri
+    };
+
+    // We aren't actually interested in reading messages, so use a blank converter option
+    rosbag2_cpp::ConverterOptions blank_converter_options {};
+    ROSBAG2_CPP_LOG_INFO_STREAM("Preparing to open...");
+    bag_reader->open(temp_so, blank_converter_options);
+    ROSBAG2_CPP_LOG_INFO_STREAM("Opened");
+    // Get metadata from the bag file
+    auto temp_metadata = bag_reader->get_metadata();
+
+    if (temp_metadata.starting_time < metadata_.starting_time) {
+      metadata_.starting_time = temp_metadata.starting_time;
+    }
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //  "Current duration: " + std::to_string(metadata_.duration.count()));
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //   "Incoming duration: " + std::to_string(temp_metadata.duration.count()));
+    metadata_.duration += temp_metadata.duration;
+    ROSBAG2_CPP_LOG_INFO_STREAM("New duration: " + std::to_string(metadata_.duration.count()));
+    metadata_.message_count += temp_metadata.message_count;
+
+    // Add the topic metadata
+    for (const auto & topic : temp_metadata.topics_with_message_count) {
+      auto found_topic = std::find_if(
+        metadata_.topics_with_message_count.begin(),
+        metadata_.topics_with_message_count.end(),
+        [&topic](const rosbag2_storage::TopicInformation & agg_topic)
+        {return topic.topic_metadata.name == agg_topic.topic_metadata.name;});
+      if (found_topic == metadata_.topics_with_message_count.end()) {
+        // It's a new topic. Add it.
+        metadata_.topics_with_message_count.emplace_back(topic);
+      } else {
+        // Merge in the new information
+        found_topic->message_count += topic.message_count;
+        if (topic.topic_metadata.offered_qos_profiles != "") {
+          found_topic->topic_metadata.offered_qos_profiles =
+            topic.topic_metadata.offered_qos_profiles;
+        }
+        if (topic.topic_metadata.serialization_format != "") {
+          found_topic->topic_metadata.serialization_format =
+            topic.topic_metadata.serialization_format;
+        }
+        if (topic.topic_metadata.type != "") {
+          found_topic->topic_metadata.type = topic.topic_metadata.type;
+        }
+      }
+    }
+
+    ROSBAG2_CPP_LOG_INFO("Closing database");
+    bag_reader->reset();   // Close the reader's storage
+  }
+  bag_reader.reset();   // Delete the reader
```

Comment:
This call is redundant - by using a smart pointer you have already guaranteed that the object will be deleted when the scope exits